btx/processing/pipca.py

```diff
@@ -0,0 +1,1011 @@
+            iPCA principal axes from model.
+        S_tot : ndarray, shape (1 x q)
+            iPCA singular values from model.
+        mu_tot : ndarray, shape (1 x d)
+            Data mean computed from all input images.
+        var_tot : ndarray, shape (1 x d)
+            Sample data variance computed from all input images.
+        """
+        if self.rank == 0:
+            U_tot = np.empty(self.num_features * self.num_components)
+            mu_tot = np.empty((self.num_features, 1))
+            var_tot = np.empty((self.num_features, 1))
+        else:
+            U_tot, mu_tot, var_tot = None, None, None
+
+        start_indices = self.split_indices[:-1]
+
+        self.comm.Gatherv(
+            self.U.flatten(),
+            [
+                U_tot,
+                self.split_counts * self.num_components,
+                start_indices * self.num_components,
+                MPI.DOUBLE,
+            ],
+            root=0,
+        )
+
+        if self.rank == 0:
+            U_tot = np.reshape(U_tot, (self.num_features, self.num_components))
+
+        self.comm.Gatherv(
+            self.mu,
+            [
+                mu_tot,
+                self.split_counts * self.num_components,
+                start_indices,
+                MPI.DOUBLE,
+            ],
+            root=0,
+        )
+        self.comm.Gatherv(
+            self.total_variance,
+            [
+                var_tot,
+                self.split_counts * self.num_components,
+                start_indices,
+                MPI.DOUBLE,
+            ],
+            root=0,
+        )
+
+        S_tot = self.S
+
+        return U_tot, S_tot, mu_tot, var_tot
+
+    def get_loss_stats(self):
+
+        if self.rank == 0:
+            print(self.outliers)
+
+    def record_loadings(self, X):
+        """
+        Method to store all loadings, ΣV^T, from present batch using past
+        model iteration.
+
+        Parameters
+        ----------
+        X : ndarray, shape (_ x m)
+            Local subdivision of current image data batch.
+        """
+        _, m = X.shape
+        n, d = self.num_incorporated_images, self.num_features
+
+        start_indices = self.split_indices[:-1]
+
+        U, _, mu, _ = self.get_model()
+
+        if self.rank == 0:
+            X_tot = np.empty((d, m))
+        else:
+            X_tot = None
+
+        self.comm.Gatherv(
+            X.flatten(),
+            [
+                X_tot,
+                self.split_counts * m,
+                start_indices * m,
+                MPI.DOUBLE,
+            ],
+            root=0,
+        )
+
+        if self.rank == 0:
+
+            X_tot = np.reshape(X_tot, (d, m))
+            cb = X_tot - np.tile(mu, (1, m))
+
+            pcs = U.T @ cb
+            self.pc_data = (
+                np.concatenate((self.pc_data, pcs), axis=1)
+                if len(self.pc_data)
+                else pcs
+            )
+
+            pc_dist = np.linalg.norm(pcs[:5], axis=0)
```

Comment:
Should this be made an argument rather than hardcoded?

Reply:
+1

Reply:
Fixed.